samtranslator/model/sam_resources.py

```diff
@@ -2194,3 +2202,108 @@ def _construct_appsync_schema(self, api_id: Intrinsicable[str]) -> GraphQLSchema
         schema.DefinitionS3Location = self.SchemaUri
 
         return schema
+
+
+class SamGraphQLDataSource(SamResourceMacro):
+    """SAM GraphQL Data Source Macro (WIP)."""
+
+    # TODO: This currently only supports DataSource with a defined ServiceRoleArn.
+    #       The proposed connector change will come in a follow up PR.
+    resource_type = "AWS::Serverless::GraphQLDataSource"
+    property_types = {
+        "ApiId": Property(True, one_of(IS_STR, IS_DICT)),
+        "Name": Property(False, IS_STR),
+        "Type": Property(True, IS_STR),
+        "Description": PassThroughProperty(False),
+        "ServiceRoleArn": Property(False, IS_STR),
+        "Permissions": Property(False, list_of(IS_STR)),
+        "DataSourceConfig": Property(True, IS_DICT),
+    }
+
+    ApiId: Intrinsicable[str]
+    Type: str
+    DataSourceConfig: DataSourceConfig
```

Comment:
I think you want to wrap Configs in additional `dict` to make validation easier, but from UX perspective one more level doesn't make much sense 
```yaml
DataSourceConfig:
  DynamoDb:
    UseCallerCredentials: True
```
From `Type` field, we already know, which fields should be in the config. So 
```yaml
Config:
  UseCallerCredentials: True
```
It seems cleaner. And I'd call it simply `Config` because it's in `DataSource` already. If `Config` sounds too general, I guess `ConnectionConfig` would be more precise

Reply:
That's an interesting idea. What would documentation look like for the `Config` property? I'm concerned that it would make that property bloated having it support all the different datasource properties.

Reply:
I agree that adding another layer is not necessary. But I'm going to keep the configs separate. I just think the property gets too bloated and it could be more of a detriment to customers than a benefit.